### PR TITLE
AvroConfluent: skip Kafka tombstone records/skip broken records

### LINKED ISSUE #12661: Exception when reading from Kafka
Hi,

I am running a Clickhouse cluster on k8s and reading from Kafka topics. I am using Clickhouse 20.4.4 .18 and the 0.9.9 version of the Clickhouse operator.  The topics are in AvroConfluent format and everything worked well until I started having an exception for on eof the topics where the complaint is :"Invalid name space"
The name of namespace is something like xxx-yyyy.prod. 

Do you have any idea what this can be about? 
Many thanks!

### PR DESCRIPTION
**Pull request by oandrew**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
AvroConfluent: Skip Kafka tombstone records
AvroConfluent: Support skipping broken records
...

Detailed description / Documentation draft:
Kafka tombstone records have null value and should be skipped instead of throwing an error.

Fixes: #12661 - you can skip broken (bad schema) messages. 
...

By adding documentation, you'll allow users to try your new feature immediately, not when someone else will have time to document it later. Documentation is necessary for all features that affect user experience in any way. You can add brief documentation draft above, or add documentation right into your patch as Markdown files in [docs](https://github.com/ClickHouse/ClickHouse/tree/master/docs) folder.

If you are doing this for the first time, it's recommended to read the lightweight [Contributing to ClickHouse Documentation](https://github.com/ClickHouse/ClickHouse/tree/master/docs/README.md) guide first.

### CODE REVIEW

File: tests/integration/test_storage_kafka/test.py [hunk 1]
```diff
@@ -407,7 +435,21 @@ def test_kafka_formats(kafka_cluster):
...
-        # TODO: test for AvroConfluence
+        'AvroConfluent' : {
+            'data_sample': [
+                avro_confluent_message(cluster.schema_registry_client, {'id':0L,'blockNo':0,'val1':unicode('AM'),'val2':0.5,"val3":1}),
+
+                ''.join(map(lambda id: avro_confluent_message(cluster.schema_registry_client, {'id':id,'blockNo':0,'val1':unicode('AM'),'val2':0.5,"val3":1}), range(1,16))),
+
+                avro_confluent_message(cluster.schema_registry_client, {'id':0L,'blockNo':0,'val1':unicode('AM'),'val2':0.5,"val3":1}),
+
+                '',
```
Comment (oandrew) [inline tests/integration/test_storage_kafka/test.py:446]: currently 'empty value' will not fail the test because it is the last message on kafka topic (the error would happen after we have received all expected rows).


we should probably send 'empty value' first  + shift offsets in `expected` TSV by +1.

**Comment by azat** [comment]
@oandrew is this PR fixes #12661 ? If yes, can you please add `Fixes: #12661` into the description?

File: tests/integration/test_storage_kafka/test.py (hunk 1 above)
Comment (azat) [reply tests/integration/test_storage_kafka/test.py:446]: I don't understand why this test should not fail, but I've verified this w/o this PR and indeed, it does not fail, however it should (otherwise what is the purpose of the test?)

**Comment by oandrew** [comment]
> @oandrew is this PR fixes #12661 ? If yes, can you please add `Fixes: #12661` into the description?



> @oandrew is this PR fixes #12661 ? If yes, can you please add `Fixes: #12661` into the description?

Nope, skipping broken messages would be a separate PR.

File: tests/integration/test_storage_kafka/test.py (hunk 1 above)
Comment (oandrew) [reply tests/integration/test_storage_kafka/test.py]: Updated the code to test empty messages for formats that support it

File: tests/integration/test_storage_kafka/test.py [hunk 2]
```diff
@@ -452,10 +475,15 @@ def test_kafka_formats(kafka_cluster):
         # },
     }
```
Comment (azat) [inline tests/integration/test_storage_kafka/test.py:476]: Ugh, this should be rewritten as @pytest.mark.parametrize, @oandrew do you have passion to do this? (this will make test for each format as a separate)

Comment (oandrew) [reply tests/integration/test_storage_kafka/test.py:476]: Maybe with next PR.

**Review by azat — COMMENTED** [review]
No test for tombstone records, apart from this LGTM

<details>

Error with stack trace for Avro format before this PR:

```
2020.08.03 11:05:15.046158 [ 18 ] {} <Error> void DB::StorageKafka::threadFunc(): Code: 33, e.displayText() = DB::Exception: Cannot read all data. Bytes read: 0. Bytes expected: 1., Stack trace (when copying this message, always include the lines below):

0. /work3/azat/ch/clickhouse/.cmake/../contrib/poco/Foundation/src/Exception.cpp:27: Poco::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0x120158cc in /usr/bin/clickhouse
1. /work3/azat/ch/clickhouse/.cmake/../src/Common/Exception.cpp:37: DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int) @ 0xa34d029 in /usr/bin/clickhouse
2. /work3/azat/ch/clickhouse/.cmake/../contrib/libcxx/include/string:2134: DB::ReadBuffer::readStrict(char*, unsigned long) @ 0xa38d267 in /usr/bin/clickhouse
3. /work3/azat/ch/clickhouse/.cmake/../src/IO/ReadHelpers.h:107: DB::AvroConfluentRowInputFormat::readRow(std::__1::vector<COW<DB::IColumn>::mutable_ptr<DB::IColumn>, std::__1::allocator<COW<DB::IColumn>::mutable_ptr<DB::IColumn> > >&, DB::RowReadExtension&) @ 0xf96fa76 in /usr/bin/clickhouse
4. /work3/azat/ch/clickhouse/.cmake/../src/Processors/Formats/IRowInputFormat.cpp:64: DB::IRowInputFormat::generate() @ 0xf98bbb9 in /usr/bin/clickhouse
5. /work3/azat/ch/clickhouse/.cmake/../contrib/libcxx/include/vector:1003: DB::ISource::work() @ 0xf912977 in /usr/bin/clickhouse
6. /work3/azat/ch/clickhouse/.cmake/../src/Storages/Kafka/KafkaBlockInputStream.cpp:103: DB::KafkaBlockInputStream::readImpl()::'lambda'()::operator()() const @ 0xf58dee5 in /usr/bin/clickhouse
7. /work3/azat/ch/clickhouse/.cmake/../src/Storages/Kafka/KafkaBlockInputStream.cpp:139: DB::KafkaBlockInputStream::readImpl() @ 0xf58f124 in /usr/bin/clickhouse
8. /work3/azat/ch/clickhouse/.cmake/../contrib/libcxx/include/vector:1003: DB::IBlockInputStream::read() @ 0xec1895d in /usr/bin/clickhouse
9. /work3/azat/ch/clickhouse/.cmake/../src/DataStreams/copyData.cpp:26: DB::copyData(DB::IBlockInputStream&, DB::IBlockOutputStream&, std::__1::atomic<bool>*) @ 0xec39d0a in /usr/bin/clickhouse
10. /work3/azat/ch/clickhouse/.cmake/../contrib/libcxx/include/vector:1516: DB::StorageKafka::streamToViews() @ 0xf56b83d in /usr/bin/clickhouse
11. /work3/azat/ch/clickhouse/.cmake/../src/Storages/Kafka/StorageKafka.cpp:487: DB::StorageKafka::threadFunc() @ 0xf56c238 in /usr/bin/clickhouse
12. /work3/azat/ch/clickhouse/.cmake/../src/Common/Stopwatch.h:12: DB::BackgroundSchedulePoolTaskInfo::execute() @ 0xed9f262 in /usr/bin/clickhouse
13. /work3/azat/ch/clickhouse/.cmake/../contrib/poco/Foundation/include/Poco/RefCountedObject.h:82: DB::BackgroundSchedulePool::threadFunction() @ 0xed9f68a in /usr/bin/clickhouse
14. /work3/azat/ch/clickhouse/.cmake/../src/Common/ThreadPool.h:170: ThreadFromGlobalPool::ThreadFromGlobalPool<DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, unsigned long, char const*)::'lambda'()>(DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, unsigned long, unsigned long, char const*)::'lambda'()&&)::'lambda'()::operator()() const @ 0xed9f78f in /usr/bin/clickhouse
15. /work3/azat/ch/clickhouse/.cmake/../contrib/libcxx/include/atomic:856: ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xa377ead in /usr/bin/clickhouse
16. /work3/azat/ch/clickhouse/.cmake/../contrib/libcxx/include/memory:2615: void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, int, std::__1::optional<unsigned long>)::'lambda1'()> >(void*) @ 0xa3765ef in /usr/bin/clickhouse
17. start_thread @ 0x9669 in /usr/lib/x86_64-linux-gnu/libpthread-2.30.so
18. __clone @ 0x122323 in /usr/lib/x86_64-linux-gnu/libc-2.30.so
 (version 20.7.1.1)
```

</details>

**Comment by azat** [comment]
And FWIW librdkafka has been upgraded to 1.5 (in #12991), worth rebasing against upstream/master or merging it.

**Comment by oandrew** [comment]
@azat tombstone  == empty message so that is covered. 

There is no test for broken message though

**Comment by azat** [comment]
>@azat tombstone == empty message so that is covered.
>There is no test for broken message though

Egh, sorry mix up those two, how about test for broken message then? (does not looks complex at first glance)

**Comment by oandrew** [comment]
Actually it's more complex than I thought.
I tried registering a schema  with bad name (for #12661 ) but none of the tools allow you to do that (they all have validations). 
So  we would need to mock the schema registry inside the test. 
I did verify it locally though.

On top of that I noticed that IRowInputFormat skips only specific kinds of errors:
https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/Formats/IRowInputFormat.cpp#L26

For a format like AvroConfluent you can think of different kinds of broken messages:
invalid wire format (bad magic byte, EOF) - can't skip to next avro message since we don't know where it starts
schema found but is invalid (#12661) or incompatible with table columns - can skip
error during "parsing" (e.g. string as uuid fails) - can skip 


So I think empty messages/bad messages  should be handled by KafkaBlockInputStream instead of the InputFormat itself.

TL;DR
I would merge this PR as-is and revisit this later.  Adding test for broken schema is lots of work.

**Review by azat — APPROVED** [review]

**Comment by akuzm** [comment]
Functional stateless test -- 00834_kill_mutation, known problem
Performance test -- long synthetic_hardware_benchmark test, also known problem

**Comment by gervarela** [comment]
> Actually it's more complex than I thought.
> I tried registering a schema with bad name (for #12661 ) but none of the tools allow you to do that (they all have validations).
> So we would need to mock the schema registry inside the test.
> I did verify it locally though.
> 
> On top of that I noticed that IRowInputFormat skips only specific kinds of errors:
> https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/Formats/IRowInputFormat.cpp#L26
> 
> For a format like AvroConfluent you can think of different kinds of broken messages:
> invalid wire format (bad magic byte, EOF) - can't skip to next avro message since we don't know where it starts
> schema found but is invalid (#12661) - can't skip
> schema found and is valid but incompatible with table columns - can skip
> error during "parsing" (e.g. string as uuid fails) - can skip
> 
> So I think empty messages/bad messages should be handled by KafkaBlockInputStream because it can simply skip to next kafka message
> 
> TL;DR
> I would merge this PR as-is and revisit this later. Adding test for broken schema is lots of work.

Hi all,

Just found this "pull request" while investigating if the option 'kafka_skip_broken_messages' works with AvroConfluent format. 

We have found that writing a message with a wrong schema to a topic (I mean, a valid schema, but no the one expected for the topic) ends up with the error: " DB::Exception: Field xxxxx not found in Avro schema: While executing SourceFromInputStream" and the processing of the topic gest blocked, no new events are processed after the wrong one enters the topic.

¿Does this solution covers this use case? ¿Does it actually uses the 'kafka_skip_broken_messages' option?.

Regarding the cited message, this particular error may be considered "INCORRECT_DATA" according to IRowInputFormat.

Thank you.

**Comment by oandrew** [comment]
@gervarela 
Setting  `input_format_avro_allow_missing_fields`  should help you
https://github.com/ClickHouse/ClickHouse/pull/12007

**Comment by gervarela** [comment]
@oandrew thank you for your support

mmm, just tried it and, well it can be a workaround. It works in the sense that it now passes the "Avro deserialization" step, but it gets continue to get stuck due to an error at the materialized view level because, obviusly, the record doesn't have any (or most) ot the attributes required by schema, so they are null.

This solution will require to rethink the materialized view so that it can deal with "very bad data" (a lot of nulls). We alredy do it with some field, but for others we consider them mandatory and expect the schema to enfornce them.

So that I think that this is a different use case where the ideal solution would be to just ignore any message that does not comply with the expected schema. In fact, for this very specific case, may the simple solution is to check that the schema ID of the message matches the schema ID of the topic, and discard it otherwise?
